<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Non-stationarity</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/clipboard-2.0.6/clipboard.min.js"></script>
<link href="site_libs/xaringanExtra-clipboard-0.2.6/xaringanExtra-clipboard.css" rel="stylesheet" />
<script src="site_libs/xaringanExtra-clipboard-0.2.6/xaringanExtra-clipboard.js"></script>
<script>window.xaringanExtraClipboard(null, {"button":"<i class=\"fa-solid fa-clipboard\" style=\"color: #00008B\"><\/i>","success":"<i class=\"fa fa-check\" style=\"color: #90BE6D\"><\/i>","error":"<i class=\"fa fa-times-circle\" style=\"color: #F94144\"><\/i>"})</script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GWMF</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="preprocessing.html">Preprocessing</a>
</li>
<li>
  <a href="modeling.html">Modeling</a>
</li>
<li>
  <a href="non_stationarity.html">Non-stationarity</a>
</li>
<li>
  <a href="basis_functions.html">Basis Functions</a>
</li>
<li>
  <a href="approximation.html">Approximation</a>
</li>
<li>
  <a href="numerical_error.html">Numerical Error</a>
</li>
<li>
  <a href="convergence_rates.html">Convergence Rates</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/leninrafaelrierasegura/">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Non-stationarity</h1>
<h4 class="date">Created: 05-07-2024. Last modified: 11-07-2024.</h4>

</div>


<style type="text/css">
body .main-container {
  max-width: 100% !important;
  width: 100% !important;
}
body {
  max-width: 100% !important;
}

body, td {
   font-size: 16px;
}
code.r{
  font-size: 14px;
}
pre {
  font-size: 14px
}
</style>
<p>Go back to the <a href="about.html">About page</a>.</p>
<p>Let us set some global options for all code chunks in this
document.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  <span class="at">message =</span> <span class="cn">FALSE</span>,    <span class="co"># Disable messages printed by R code chunks</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  <span class="at">warning =</span> <span class="cn">FALSE</span>,    <span class="co"># Disable warnings printed by R code chunks</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  <span class="at">echo =</span> <span class="cn">TRUE</span>,        <span class="co"># Show R code within code chunks in output</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  <span class="at">include =</span> <span class="cn">TRUE</span>,     <span class="co"># Include both R code and its results in output</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  <span class="at">eval =</span> <span class="cn">TRUE</span>,       <span class="co"># Evaluate R code chunks</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="at">cache =</span> <span class="cn">FALSE</span>,       <span class="co"># Enable caching of R code chunks for faster rendering</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="at">fig.align =</span> <span class="st">&quot;center&quot;</span>,</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="at">out.width =</span> <span class="st">&quot;100%&quot;</span>,</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  <span class="at">retina =</span> <span class="dv">2</span>,</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="at">error =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>  <span class="at">collapse =</span> <span class="cn">FALSE</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>)</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1982</span>)</span></code></pre></div>
<div id="import-libraries" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Import libraries</h1>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(INLA)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">library</span>(inlabru)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">library</span>(rSPDE)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">library</span>(MetricGraph)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">library</span>(plotly)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="fu">library</span>(rmarkdown)</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="fu">library</span>(grateful) <span class="co"># Cite all loaded packages</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>()) <span class="co"># Clear the workspace</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1982</span>) <span class="co"># Set seed for reproducibility</span></span></code></pre></div>
</div>
<div id="build-the-graph" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Build the graph</h1>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>edges <span class="ot">&lt;-</span> <span class="fu">logo_lines</span>() <span class="co"># This is a function from MetricGraph package that returns a list of edges</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>logo_graph <span class="ot">&lt;-</span> metric_graph<span class="sc">$</span><span class="fu">new</span>(<span class="at">edges =</span> edges) <span class="co"># Create a new graph object</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#logo_graph$prune_vertices()</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">build_mesh</span>(<span class="at">h =</span> <span class="fl">0.05</span>) <span class="co"># Build the mesh</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># Plot graph with mesh</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">plot</span>(<span class="at">mesh =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mesh&quot;</span>) <span class="sc">+</span> </span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-5-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="simulate-a-whittlematérn-field" class="section level1"
number="3">
<h1><span class="header-section-number">3</span> Simulate a
Whittle–Matérn field</h1>
<p>Below we simulate a Whittle–Matérn field <span
class="math inline">\(\dot{u}\)</span> (<code>cov</code>) from the
model</p>
<p><span class="math display">\[\begin{equation}
\label{SPDE}\tag{1}
(\kappa^2-\Delta)^{\alpha/2}\tau \dot{u} = \mathcal{W}.
\end{equation}\]</span></p>
<p>We use the <code>matern</code> parameterization, where the parameters
are <span class="math inline">\(\nu\)</span> (<code>nu</code>), <span
class="math inline">\(\sigma\)</span> (<code>sigma</code>), and <span
class="math inline">\(\rho\)</span> (<code>range</code>), and are
related to the <code>spde</code> parameterization <span
class="math inline">\(\eqref{SPDE}\)</span> as follows.</p>
<ul>
<li><span class="math inline">\(\alpha = \nu + 1/2\)</span>,</li>
<li><span class="math inline">\(\rho =
\dfrac{\sqrt{8\nu}}{\kappa}\)</span>,</li>
<li><span class="math inline">\(\tau^2 = \dfrac{\Gamma(\nu)}{\sigma^2
\kappa^{2\nu}(4\pi)^{d/2}\Gamma(\nu + d/2)}\)</span>.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Set the parameters</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>sigma <span class="ot">&lt;-</span> <span class="fl">0.1</span> <span class="co"># Standard deviation</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>range <span class="ot">&lt;-</span> <span class="dv">10</span> <span class="co"># Range</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>nu <span class="ot">&lt;-</span> <span class="fl">1.5</span> <span class="co"># Smoothness parameter</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>theta <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="sc">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span>) <span class="co"># Log-regression coefficients</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>nu_ns <span class="ot">&lt;-</span> <span class="fl">1.5</span> <span class="co"># Smoothness parameter for non-stationarity</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>cor_threshold <span class="ot">&lt;-</span> <span class="fl">0.1</span> <span class="co"># Define correlation threshold</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>sigma.e <span class="ot">&lt;-</span> <span class="fl">0.1</span> <span class="co"># Standard deviation of the error term</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Construct the approximation</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>op <span class="ot">&lt;-</span> <span class="fu">matern.operators</span>(<span class="at">nu =</span> nu, </span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>                       <span class="at">sigma =</span> sigma,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>                       <span class="at">range =</span> range,  </span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>                       <span class="at">parameterization =</span> <span class="st">&quot;matern&quot;</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>                       <span class="at">graph =</span> logo_graph)    </span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co"># Simulate the field</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>cov <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">simulate</span>(op))</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co"># Plot the field</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">plot_function</span>(<span class="at">X =</span> cov, <span class="at">vertex_size =</span> <span class="dv">0</span>, <span class="at">plotly =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Simulated field&quot;</span>) <span class="sc">+</span> </span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-7-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div
id="build-a-non-stationary-covariance-function-for-a-generalized-whittlematérn-field"
class="section level1" number="4">
<h1><span class="header-section-number">4</span> Build a non-stationary
covariance function for a generalized Whittle–Matérn field</h1>
<p>We use the simulated field to build a non-stationary covariance
function for a generalized Whittle–Matérn field <span
class="math inline">\(u\)</span>. That is, we consider the model</p>
<p><span class="math display">\[\begin{equation}
\label{NS-SPDE}\tag{2}
(\kappa^2(s)-\Delta)^{\alpha/2}\tau(s) u = \mathcal{W},
\end{equation}\]</span></p>
<p>with non-stationary parameters (in the <code>matern</code>
parameterization) given by</p>
<p><span class="math display">\[\begin{equation}
\label{logregressions}
    \begin{aligned}
    \log(\sigma(s)) &amp;= \theta_1 + \theta_3 \text{cov}(s),\\
    \log(\rho(s)) &amp;= \theta_2 + \theta_4 \text{cov}(s).
\end{aligned}
\end{equation}\]</span></p>
<p>Take into account that <code>theta</code> <span
class="math inline">\(= (\theta_1, \theta_2, \theta_3,
\theta_4)\)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Non-stationary parameters</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>B.sigma <span class="ot">=</span> <span class="fu">cbind</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, cov, <span class="dv">0</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>B.range <span class="ot">=</span> <span class="fu">cbind</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, cov)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co"># Construct the approximation </span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>ns_op <span class="ot">&lt;-</span> rSPDE<span class="sc">::</span><span class="fu">spde.matern.operators</span>(<span class="at">graph =</span> logo_graph,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>                                                <span class="at">parameterization =</span> <span class="st">&quot;matern&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>                                                <span class="at">B.sigma =</span> B.sigma,</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>                                                <span class="at">B.range =</span> B.range,</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>                                                <span class="at">theta =</span> theta,</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>                                                <span class="at">nu =</span> nu_ns)</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co"># Covariance matrix</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>est_cov_matrix <span class="ot">&lt;-</span> ns_op<span class="sc">$</span><span class="fu">covariance_mesh</span>()</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co"># Precision matrix</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>Q <span class="ot">&lt;-</span> <span class="fu">precision</span>(ns_op)</span></code></pre></div>
<div id="plot-the-covariance-between-a-point-and-all-other-points"
class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Plot the covariance
between a point and all other points</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>rownumber <span class="ot">=</span> <span class="dv">300</span> <span class="co"># Choose a row number, any berween 1 and dim(est_cov_matrix)[1]</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>rowfromcov <span class="ot">=</span> est_cov_matrix[rownumber, ] <span class="co"># Get a row from the covariance matrix</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>rowfromQ <span class="ot">=</span> <span class="fu">solve</span>(Q, <span class="fu">replace</span>(<span class="fu">numeric</span>(<span class="fu">dim</span>(Q)[<span class="dv">1</span>]), rownumber, <span class="dv">1</span>)) <span class="co"># Another way to get a row from the covariance matrix</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">sum</span>(rowfromcov <span class="sc">-</span> rowfromQ) <span class="co"># Check they are the same</span></span></code></pre></div>
<pre><code>## [1] -6.063467e-10</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Plot the covariance between a point and all other points</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">plot_function</span>(<span class="at">X =</span> rowfromQ, <span class="at">vertex_size =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Covariance between a point and all other points&quot;</span>) <span class="sc">+</span> </span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-10-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div
id="compute-the-practical-correlation-range-and-the-standard-deviation"
class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Compute the practical
correlation range and the standard deviation</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Build the models for sigma and range</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>model_for_sigma <span class="ot">&lt;-</span> <span class="fu">exp</span>(B.sigma[,<span class="sc">-</span><span class="dv">1</span>]<span class="sc">%*%</span>theta)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>model_for_range <span class="ot">&lt;-</span> <span class="fu">exp</span>(B.range[,<span class="sc">-</span><span class="dv">1</span>]<span class="sc">%*%</span>theta)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># Compute distance matrix on the mesh</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">compute_geodist_mesh</span>()</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co"># Get the distance matrix</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>dist_matrix <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span>mesh<span class="sc">$</span>geo_dist</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co"># Initialize vector to store the practical correlation range</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>est_range <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">dim</span>(est_cov_matrix)[<span class="dv">1</span>])</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>  </span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co"># Compute the practical correlation range</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(est_cov_matrix)[<span class="dv">1</span>]) {</span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>  rowi <span class="ot">&lt;-</span> est_cov_matrix[i,]</span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>  sorted <span class="ot">&lt;-</span> <span class="fu">sort</span>(rowi, <span class="at">decreasing =</span> <span class="cn">TRUE</span>, <span class="at">index.return =</span> <span class="cn">TRUE</span>)</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>  index_cor_smaller_pcr <span class="ot">&lt;-</span> <span class="fu">which</span>(sorted<span class="sc">$</span>x <span class="sc">&lt;</span> cor_threshold)[<span class="dv">1</span>]</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>  pos_dist_matrix <span class="ot">&lt;-</span> sorted<span class="sc">$</span>ix[index_cor_smaller_pcr]</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a>  est_range[i] <span class="ot">&lt;-</span> dist_matrix[i, pos_dist_matrix]</span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>}</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="co"># Compute the standard deviation</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a>est_sigma <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">diag</span>(INLA<span class="sc">::</span><span class="fu">inla.qinv</span>(Q))) <span class="co"># It can also be computed as sqrt(Matrix::diag(est_cov_matrix))</span></span></code></pre></div>
</div>
<div
id="plot-the-practical-correlation-range-and-the-standard-deviation"
class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Plot the practical
correlation range and the standard deviation</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Extract the mesh locations in Euclidean coordinates</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>xypoints <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span>mesh<span class="sc">$</span>V</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co"># Define data frames for plotting</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>sigma_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> xypoints[,<span class="dv">1</span>], <span class="at">y =</span> xypoints[,<span class="dv">2</span>], <span class="at">z =</span> model_for_sigma, <span class="at">.group =</span> <span class="dv">1</span>)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>range_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> xypoints[,<span class="dv">1</span>], <span class="at">y =</span> xypoints[,<span class="dv">2</span>], <span class="at">z =</span> model_for_range, <span class="at">.group =</span> <span class="dv">2</span>)</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>sd_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> xypoints[,<span class="dv">1</span>], <span class="at">y =</span> xypoints[,<span class="dv">2</span>], <span class="at">z =</span> est_sigma, <span class="at">.group =</span> <span class="dv">3</span>)</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>rho_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> xypoints[,<span class="dv">1</span>], <span class="at">y =</span> xypoints[,<span class="dv">2</span>], <span class="at">z =</span> est_range, <span class="at">.group =</span> <span class="dv">4</span>)</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>data_param_cov <span class="ot">&lt;-</span> <span class="fu">rbind</span>(sigma_df, range_df, sd_df, rho_df)</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>group_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Model for sigma&quot;</span>, <span class="st">&quot;Model for range&quot;</span>, <span class="st">&quot;Standard deviation&quot;</span>, <span class="st">&quot;Practical correlation range&quot;</span>)</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>min_x <span class="ot">&lt;-</span> <span class="fu">min</span>(data_param_cov<span class="sc">$</span>x) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>)</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>max_x <span class="ot">&lt;-</span> <span class="fu">max</span>(data_param_cov<span class="sc">$</span>x) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>)</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>min_y <span class="ot">&lt;-</span> <span class="fu">min</span>(data_param_cov<span class="sc">$</span>y) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>)</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>max_y <span class="ot">&lt;-</span> <span class="fu">max</span>(data_param_cov<span class="sc">$</span>y) <span class="sc">|&gt;</span> <span class="fu">round</span>(<span class="dv">2</span>)</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="co"># Create a base plot</span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>base_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> sigma_df, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y), <span class="at">color =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">shape =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">labels =</span> <span class="cf">function</span>(x) <span class="fu">paste0</span>(x), <span class="at">name =</span> <span class="st">&quot;x&quot;</span>, <span class="at">breaks =</span> <span class="fu">seq</span>(min_x, max_x, <span class="at">by =</span> <span class="dv">2</span>)) <span class="sc">+</span>  </span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">labels =</span> <span class="cf">function</span>(y) <span class="fu">paste0</span>(y), <span class="at">name =</span> <span class="st">&quot;y&quot;</span>, <span class="at">breaks =</span> <span class="fu">seq</span>(min_y, max_y, <span class="at">by =</span> <span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a><span class="co"># Create individual plots for each group and overlay data points</span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a>group_plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">unique</span>(data_param_cov<span class="sc">$</span>.group), <span class="cf">function</span>(gr) {</span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> <span class="fu">subset</span>(data_param_cov, .group <span class="sc">==</span> gr), <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">color =</span> z), <span class="at">size =</span> <span class="dv">1</span>)</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a>})</span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="co"># Combine the base plot and the group plots</span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a>combined_plot <span class="ot">&lt;-</span> base_plot <span class="sc">+</span> </span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> .group, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labeller =</span> <span class="fu">labeller</span>(<span class="at">.group =</span> <span class="cf">function</span>(variable, value) {</span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a>    <span class="fu">return</span>(group_labels[value])</span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a>  })) <span class="sc">+</span>  <span class="co"># Specify custom labels for facet_wrap</span></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>  group_plots <span class="sc">+</span></span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>  <span class="fu">scale_color_gradient</span>(<span class="at">name =</span> <span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">&quot;&quot;</span>) <span class="sc">+</span></span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">option =</span> <span class="st">&quot;D&quot;</span>)</span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a></span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a><span class="co"># Print the combined plot</span></span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a><span class="fu">print</span>(combined_plot)</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-12-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="simulation" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> Simulation</h2>
</div>
<div id="simulate-and-plot-the-non-stationary-field"
class="section level2" number="4.5">
<h2><span class="header-section-number">4.5</span> Simulate and plot the
non-stationary field</h2>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># Simulate the non-stationary field</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>u_non_stat <span class="ot">&lt;-</span> <span class="fu">simulate</span>(ns_op)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co"># Plot the non-stationary field</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">plot_function</span>(<span class="at">X =</span> u_non_stat, <span class="at">vertex_size =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Simulated non-stationary field&quot;</span>) <span class="sc">+</span> </span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-13-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="simulate-observations" class="section level2" number="4.6">
<h2><span class="header-section-number">4.6</span> Simulate
observations</h2>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Number of observations per edge</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>obs.per.edge <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>obs.loc <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>logo_graph<span class="sc">$</span>nE) {</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>  obs.loc <span class="ot">&lt;-</span> <span class="fu">rbind</span>(obs.loc,</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>                   <span class="fu">cbind</span>(<span class="fu">rep</span>(i,obs.per.edge), <span class="fu">runif</span>(obs.per.edge)))</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>}</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>n.obs <span class="ot">&lt;-</span> obs.per.edge<span class="sc">*</span>logo_graph<span class="sc">$</span>nE</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="co"># Compute the FEM basis functions on the observation locations</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>A <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span><span class="fu">fem_basis</span>(obs.loc)</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co"># Define two covariates</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">sin</span>(obs.loc[,<span class="dv">1</span>])</span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">sin</span>(obs.loc[,<span class="dv">2</span>])</span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="co"># Simulate observations</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>Y <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">+</span> <span class="dv">2</span><span class="sc">*</span>x1 <span class="sc">-</span> <span class="dv">3</span><span class="sc">*</span>x2 <span class="sc">+</span> <span class="fu">as.vector</span>(A <span class="sc">%*%</span> u_non_stat <span class="sc">+</span> sigma.e <span class="sc">*</span> <span class="fu">rnorm</span>(n.obs))</span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="co"># Create a data frame with the observations</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a>df_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">y =</span> Y, <span class="at">edge_number =</span> obs.loc[,<span class="dv">1</span>],</span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a>                        <span class="at">distance_on_edge =</span> obs.loc[,<span class="dv">2</span>],</span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a>                        <span class="at">x1 =</span> x1, <span class="at">x2 =</span> x2)</span></code></pre></div>
</div>
<div id="compare-the-data-before-and-after-adding-the-observations"
class="section level2" number="4.7">
<h2><span class="header-section-number">4.7</span> Compare the data
before and after adding the observations</h2>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Add the observations to the graph</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">add_observations</span>(<span class="at">data =</span> df_data, <span class="at">normalized =</span> <span class="cn">TRUE</span>, <span class="at">clear_obs =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>data_after_added <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span><span class="fu">get_data</span>() <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">something =</span> <span class="dv">1</span>)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co"># Add the observations to the graph</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">add_observations</span>(<span class="at">data =</span> data_after_added, <span class="at">normalized =</span> <span class="cn">TRUE</span>, <span class="at">clear_obs =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>data_after_added2 <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span><span class="fu">get_data</span>()</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="co"># Get edge_number and distance_on_edge to compare</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>toy1 <span class="ot">&lt;-</span> df_data <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(edge_number, distance_on_edge)</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>toy2 <span class="ot">&lt;-</span> data_after_added <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">distance_on_edge =</span> .distance_on_edge, <span class="at">edge_number =</span> .edge_number) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(edge_number, distance_on_edge)</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a>toy3 <span class="ot">&lt;-</span> data_after_added2 <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">distance_on_edge =</span> .distance_on_edge, <span class="at">edge_number =</span> .edge_number) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(edge_number, distance_on_edge)</span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="co"># Compare them</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="fu">identical</span>(toy1, toy2)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">identical</span>(toy2, toy3)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="add-and-plot-the-simulated-data" class="section level2"
number="4.8">
<h2><span class="header-section-number">4.8</span> Add and plot the
simulated data</h2>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># Add the observations to the graph</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">add_observations</span>(<span class="at">data =</span> df_data, <span class="at">normalized =</span> <span class="cn">TRUE</span>, <span class="at">clear_obs =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co"># Get the data</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>data <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span><span class="fu">get_data</span>()</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="co"># Plot the simulated observations</span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a>logo_graph<span class="sc">$</span><span class="fu">plot</span>(<span class="at">data =</span> <span class="st">&quot;y&quot;</span>, <span class="at">vertex_size =</span> <span class="dv">0</span>) <span class="sc">+</span> </span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Simulated data&quot;</span>) <span class="sc">+</span> </span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-16-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="compute-the-distance-matrix-and-create-the-groups"
class="section level2" number="4.9">
<h2><span class="header-section-number">4.9</span> Compute the distance
matrix and create the groups</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># Define aux data frame to compute the distance matrix</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>aux <span class="ot">&lt;-</span> data <span class="sc">|&gt;</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">distance_on_edge =</span> .distance_on_edge, <span class="at">edge_number =</span> .edge_number) <span class="sc">|&gt;</span> <span class="co"># Rename the variables (because sf_graph$compute_geodist_PtE() requires so)</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span> <span class="co"># Transform to a data frame (i.e., remove the metric_graph class)</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(edge_number, distance_on_edge)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co"># Compute the distance matrix</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>distmatrix <span class="ot">&lt;-</span> logo_graph<span class="sc">$</span><span class="fu">compute_geodist_PtE</span>(<span class="at">PtE =</span> aux,</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>                                             <span class="at">normalized =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>                                             <span class="at">include_vertices =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="co"># Define the distance vector</span></span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a>distance <span class="ot">=</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="fl">2.5</span>, <span class="at">by =</span> <span class="fl">0.25</span>)</span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a>GROUPS <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance)) {</span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a>  GROUPS[[j]] <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(aux)) {</span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a>    GROUPS[[j]][[i]] <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">as.vector</span>(distmatrix[i, ]) <span class="sc">&lt;=</span> distance[j])</span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a>  }</span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="check-groups" class="section level2" number="4.10">
<h2><span class="header-section-number">4.10</span> Check groups</h2>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>indexofinterest <span class="ot">&lt;-</span> <span class="dv">2000</span> <span class="co"># Any number between 1 and nrow(data)</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>pointofinterest <span class="ot">&lt;-</span> GROUPS[[<span class="dv">3</span>]][[indexofinterest]]</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>pointofinterest2 <span class="ot">&lt;-</span> GROUPS[[<span class="dv">11</span>]][[indexofinterest]]</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> data, <span class="fu">aes</span>(<span class="at">x =</span> .coord_x, <span class="at">y =</span> .coord_y), <span class="at">color =</span> <span class="st">&quot;black&quot;</span>) <span class="sc">+</span></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> data[pointofinterest2, ], <span class="fu">aes</span>(<span class="at">x =</span> .coord_x, <span class="at">y =</span> .coord_y), <span class="at">color =</span> <span class="st">&quot;green&quot;</span>) <span class="sc">+</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> data[pointofinterest, ], <span class="fu">aes</span>(<span class="at">x =</span> .coord_x, <span class="at">y =</span> .coord_y), <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span></span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> data[indexofinterest, ], <span class="fu">aes</span>(<span class="at">x =</span> .coord_x, <span class="at">y =</span> .coord_y), <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Groups&quot;</span>) <span class="sc">+</span> </span>
<span id="cb20-10"><a href="#cb20-10" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb20-11"><a href="#cb20-11" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>)) <span class="sc">+</span></span>
<span id="cb20-12"><a href="#cb20-12" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-18-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div
style="display: flex; to display: grid; grid-template-columns: 1fr 1fr; grid-column-gap: 10px;">
<div id="fit-the-stationary-model" class="section level2" number="4.11">
<h2><span class="header-section-number">4.11</span> Fit the stationary
model</h2>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>rspde_model_stat <span class="ot">&lt;-</span> <span class="fu">rspde.metric_graph</span>(logo_graph,</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>                                       <span class="at">parameterization =</span> <span class="st">&quot;matern&quot;</span>,</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>                                       <span class="at">nu =</span> nu_ns)</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>data_rspde_bru_stat <span class="ot">&lt;-</span> <span class="fu">graph_data_rspde</span>(rspde_model_stat,</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>                                        <span class="at">loc_name =</span> <span class="st">&quot;loc&quot;</span>)</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>cmp_stat <span class="ot">&lt;-</span> y <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a>  <span class="fu">Intercept</span>(<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>  x1 <span class="sc">+</span></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>  x2 <span class="sc">+</span></span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>  <span class="fu">field</span>(loc, <span class="at">model =</span> rspde_model_stat)</span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a></span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a>rspde_fit_stat <span class="ot">&lt;-</span></span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a>  <span class="fu">bru</span>(cmp_stat,</span>
<span id="cb21-15"><a href="#cb21-15" tabindex="-1"></a>      <span class="at">data =</span> data_rspde_bru_stat[[<span class="st">&quot;data&quot;</span>]],</span>
<span id="cb21-16"><a href="#cb21-16" tabindex="-1"></a>      <span class="at">family =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb21-17"><a href="#cb21-17" tabindex="-1"></a>      <span class="at">options =</span> <span class="fu">list</span>(<span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-18"><a href="#cb21-18" tabindex="-1"></a>  )</span>
<span id="cb21-19"><a href="#cb21-19" tabindex="-1"></a></span>
<span id="cb21-20"><a href="#cb21-20" tabindex="-1"></a><span class="fu">summary</span>(rspde_fit_stat)</span></code></pre></div>
<pre><code>## inlabru version: 2.10.1.9010
## INLA version: 24.06.27
## Components:
## Intercept: main = linear(1), group = exchangeable(1L), replicate = iid(1L)
## x1: main = linear(x1), group = exchangeable(1L), replicate = iid(1L)
## x2: main = linear(x2), group = exchangeable(1L), replicate = iid(1L)
## field: main = cgeneric(loc), group = exchangeable(1L), replicate = iid(1L)
## Likelihoods:
##   Family: &#39;gaussian&#39;
##     Data class: &#39;metric_graph_data&#39;, &#39;list&#39;
##     Predictor: y ~ .
## Time used:
##     Pre = 0.226, Running = 0.534, Post = 0.0786, Total = 0.838 
## Fixed effects:
##             mean    sd 0.025quant 0.5quant 0.975quant   mode kld
## Intercept  1.442 0.377      0.699    1.442      2.186  1.442   0
## x1         2.012 0.020      1.974    2.012      2.050  2.012   0
## x2        -2.994 0.029     -3.051   -2.994     -2.937 -2.994   0
## 
## Random effects:
##   Name     Model
##     field CGeneric
## 
## Model hyperparameters:
##                                           mean    sd 0.025quant 0.5quant
## Precision for the Gaussian observations 97.272 3.453     90.643   97.213
## Theta1 for field                         0.836 0.092      0.661    0.834
## Theta2 for field                         0.812 0.070      0.679    0.811
##                                         0.975quant   mode
## Precision for the Gaussian observations    104.237 97.102
## Theta1 for field                             1.022  0.827
## Theta2 for field                             0.953  0.805
## 
## Deviance Information Criterion (DIC) ...............: -3530.90
## Deviance Information Criterion (DIC, saturated) ....: 3042.65
## Effective number of parameters .....................: 640.17
## 
## Watanabe-Akaike information criterion (WAIC) ...: -3512.65
## Effective number of parameters .................: 540.71
## 
## Marginal log-Likelihood:  858.45 
##  is computed 
## Posterior summaries for the linear predictor and the fitted values are computed
## (Posterior marginals needs also &#39;control.compute=list(return.marginals.predictor=TRUE)&#39;)</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">rspde.result</span>(rspde_fit_stat, <span class="st">&quot;field&quot;</span>, rspde_model_stat))</span></code></pre></div>
<pre><code>##            mean       sd 0.025quant 0.5quant 0.975quant    mode
## std.dev 2.31723 0.213115    1.93878  2.30249    2.77471 2.26623
## range   2.25800 0.157284    1.97326  2.24898    2.59055 2.22718</code></pre>
</div>
<div id="fit-the-non-stationary-model" class="section level2"
number="4.12">
<h2><span class="header-section-number">4.12</span> Fit the
non-stationary model</h2>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>rspde_model_nonstat <span class="ot">&lt;-</span> <span class="fu">rspde.metric_graph</span>(logo_graph,</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>                                          <span class="at">B.sigma =</span> B.sigma,</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>                                          <span class="at">B.range =</span> B.range,</span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>                                          <span class="at">parameterization =</span> <span class="st">&quot;matern&quot;</span>,</span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>                                          <span class="at">nu =</span> nu_ns)</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>data_rspde_bru_nonstat <span class="ot">&lt;-</span> <span class="fu">graph_data_rspde</span>(rspde_model_nonstat,</span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>                                           <span class="at">loc_name =</span> <span class="st">&quot;loc&quot;</span>)</span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a>cmp_nonstat <span class="ot">&lt;-</span> y <span class="sc">~</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" tabindex="-1"></a>  <span class="fu">Intercept</span>(<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11" tabindex="-1"></a>  x1 <span class="sc">+</span></span>
<span id="cb25-12"><a href="#cb25-12" tabindex="-1"></a>  x2 <span class="sc">+</span></span>
<span id="cb25-13"><a href="#cb25-13" tabindex="-1"></a>  <span class="fu">field</span>(loc, <span class="at">model =</span> rspde_model_nonstat)</span>
<span id="cb25-14"><a href="#cb25-14" tabindex="-1"></a></span>
<span id="cb25-15"><a href="#cb25-15" tabindex="-1"></a>rspde_fit_nonstat <span class="ot">&lt;-</span></span>
<span id="cb25-16"><a href="#cb25-16" tabindex="-1"></a>  <span class="fu">bru</span>(cmp_nonstat,</span>
<span id="cb25-17"><a href="#cb25-17" tabindex="-1"></a>      <span class="at">data =</span> data_rspde_bru_nonstat[[<span class="st">&quot;data&quot;</span>]],</span>
<span id="cb25-18"><a href="#cb25-18" tabindex="-1"></a>      <span class="at">family =</span> <span class="st">&quot;gaussian&quot;</span>,</span>
<span id="cb25-19"><a href="#cb25-19" tabindex="-1"></a>      <span class="at">options =</span> <span class="fu">list</span>(<span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-20"><a href="#cb25-20" tabindex="-1"></a>  )</span>
<span id="cb25-21"><a href="#cb25-21" tabindex="-1"></a><span class="fu">summary</span>(rspde_fit_nonstat)</span></code></pre></div>
<pre><code>## inlabru version: 2.10.1.9010
## INLA version: 24.06.27
## Components:
## Intercept: main = linear(1), group = exchangeable(1L), replicate = iid(1L)
## x1: main = linear(x1), group = exchangeable(1L), replicate = iid(1L)
## x2: main = linear(x2), group = exchangeable(1L), replicate = iid(1L)
## field: main = cgeneric(loc), group = exchangeable(1L), replicate = iid(1L)
## Likelihoods:
##   Family: &#39;gaussian&#39;
##     Data class: &#39;metric_graph_data&#39;, &#39;list&#39;
##     Predictor: y ~ .
## Time used:
##     Pre = 0.143, Running = 0.678, Post = 0.104, Total = 0.925 
## Fixed effects:
##             mean    sd 0.025quant 0.5quant 0.975quant   mode kld
## Intercept  1.474 0.379      0.723    1.475      2.215  1.475   0
## x1         2.022 0.017      1.988    2.022      2.055  2.022   0
## x2        -2.976 0.025     -3.025   -2.976     -2.926 -2.976   0
## 
## Random effects:
##   Name     Model
##     field CGeneric
## 
## Model hyperparameters:
##                                           mean    sd 0.025quant 0.5quant
## Precision for the Gaussian observations 97.810 3.355     91.358   97.757
## Theta1 for field                         0.899 0.106      0.697    0.897
## Theta2 for field                         0.923 0.080      0.769    0.922
## Theta3 for field                        -1.719 2.081     -5.658   -1.770
## Theta4 for field                         3.011 1.607     -0.032    2.972
##                                         0.975quant   mode
## Precision for the Gaussian observations     104.57 97.662
## Theta1 for field                              1.11  0.889
## Theta2 for field                              1.08  0.916
## Theta3 for field                              2.53 -1.997
## Theta4 for field                              6.29  2.798
## 
## Deviance Information Criterion (DIC) ...............: -3627.17
## Deviance Information Criterion (DIC, saturated) ....: 2958.96
## Effective number of parameters .....................: 556.68
## 
## Watanabe-Akaike information criterion (WAIC) ...: -3623.07
## Effective number of parameters .................: 466.80
## 
## Marginal log-Likelihood:  1009.31 
##  is computed 
## Posterior summaries for the linear predictor and the fitted values are computed
## (Posterior marginals needs also &#39;control.compute=list(return.marginals.predictor=TRUE)&#39;)</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">rspde.result</span>(rspde_fit_nonstat, <span class="st">&quot;field&quot;</span>, rspde_model_nonstat))</span></code></pre></div>
<pre><code>##                    mean        sd 0.025quant  0.5quant 0.975quant      mode
## Theta1.matern  0.899272 0.1055690   0.697183  0.897363    1.11275  0.889067
## Theta2.matern  0.923192 0.0804035   0.768948  0.921843    1.08547  0.916006
## Theta3.matern -1.718620 2.0807000  -5.658230 -1.769820    2.52866 -1.997110
## Theta4.matern  3.011120 1.6067300  -0.032451  2.972000    6.28967  2.798480</code></pre>
</div>
</div>
<div id="cross-validation" class="section level2" number="4.13">
<h2><span class="header-section-number">4.13</span>
Cross-validation</h2>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a>mse.stat <span class="ot">&lt;-</span> mse.nonstat <span class="ot">&lt;-</span> ls.stat <span class="ot">&lt;-</span> ls.nonstat <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>,<span class="fu">length</span>(distance))</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="co"># cross-validation for-loop</span></span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(distance)) {</span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>  <span class="co"># cross-validation of the stationary model</span></span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>  cv.stat <span class="ot">&lt;-</span> <span class="fu">inla.group.cv</span>(rspde_fit_stat, <span class="at">groups =</span> GROUPS[[j]])</span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a>  <span class="co"># cross-validation of the nonstationary model</span></span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a>  cv.nonstat <span class="ot">&lt;-</span> <span class="fu">inla.group.cv</span>(rspde_fit_nonstat, <span class="at">groups =</span> GROUPS[[j]])</span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a>  <span class="co"># obtain MSE and LS</span></span>
<span id="cb29-9"><a href="#cb29-9" tabindex="-1"></a>  mse.stat[j] <span class="ot">&lt;-</span> <span class="fu">mean</span>((cv.stat<span class="sc">$</span>mean <span class="sc">-</span> data<span class="sc">$</span>y)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb29-10"><a href="#cb29-10" tabindex="-1"></a>  mse.nonstat[j] <span class="ot">&lt;-</span> <span class="fu">mean</span>((cv.nonstat<span class="sc">$</span>mean <span class="sc">-</span> data<span class="sc">$</span>y)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb29-11"><a href="#cb29-11" tabindex="-1"></a>  ls.stat[j] <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">log</span>(cv.stat<span class="sc">$</span>cv))</span>
<span id="cb29-12"><a href="#cb29-12" tabindex="-1"></a>  ls.nonstat[j] <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">log</span>(cv.nonstat<span class="sc">$</span>cv))</span>
<span id="cb29-13"><a href="#cb29-13" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="plot-cross-validation-results" class="section level2"
number="4.14">
<h2><span class="header-section-number">4.14</span> Plot
cross-validation results</h2>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="co"># Create data frames</span></span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>mse_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a>  distance,</span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a>  <span class="at">Statnu0.5 =</span> mse.stat,</span>
<span id="cb30-5"><a href="#cb30-5" tabindex="-1"></a>  <span class="at">Nonstatnu0.5 =</span> mse.nonstat</span>
<span id="cb30-6"><a href="#cb30-6" tabindex="-1"></a>)</span>
<span id="cb30-7"><a href="#cb30-7" tabindex="-1"></a></span>
<span id="cb30-8"><a href="#cb30-8" tabindex="-1"></a>ls_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb30-9"><a href="#cb30-9" tabindex="-1"></a>  distance,</span>
<span id="cb30-10"><a href="#cb30-10" tabindex="-1"></a>  <span class="at">Statnu0.5 =</span> <span class="sc">-</span>ls.stat,</span>
<span id="cb30-11"><a href="#cb30-11" tabindex="-1"></a>  <span class="at">Nonstatnu0.5 =</span> <span class="sc">-</span>ls.nonstat</span>
<span id="cb30-12"><a href="#cb30-12" tabindex="-1"></a>)</span>
<span id="cb30-13"><a href="#cb30-13" tabindex="-1"></a></span>
<span id="cb30-14"><a href="#cb30-14" tabindex="-1"></a><span class="co"># Convert to long format</span></span>
<span id="cb30-15"><a href="#cb30-15" tabindex="-1"></a>mse_long <span class="ot">&lt;-</span> mse_df <span class="sc">%&gt;%</span></span>
<span id="cb30-16"><a href="#cb30-16" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>distance, <span class="at">names_to =</span> <span class="st">&quot;nu&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;MSE&quot;</span>)</span>
<span id="cb30-17"><a href="#cb30-17" tabindex="-1"></a></span>
<span id="cb30-18"><a href="#cb30-18" tabindex="-1"></a>ls_long <span class="ot">&lt;-</span> ls_df <span class="sc">%&gt;%</span></span>
<span id="cb30-19"><a href="#cb30-19" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">-</span>distance, <span class="at">names_to =</span> <span class="st">&quot;nu&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;LogScore&quot;</span>)</span>
<span id="cb30-20"><a href="#cb30-20" tabindex="-1"></a></span>
<span id="cb30-21"><a href="#cb30-21" tabindex="-1"></a></span>
<span id="cb30-22"><a href="#cb30-22" tabindex="-1"></a><span class="co"># Update the label mappings with the new legend title</span></span>
<span id="cb30-23"><a href="#cb30-23" tabindex="-1"></a>label_mapping <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb30-24"><a href="#cb30-24" tabindex="-1"></a>  <span class="st">&quot;Statnu0.5&quot;</span> <span class="ot">=</span> <span class="st">&quot;0.5&quot;</span>, </span>
<span id="cb30-25"><a href="#cb30-25" tabindex="-1"></a>  <span class="st">&quot;Nonstatnu0.5&quot;</span> <span class="ot">=</span> <span class="st">&quot;0.5&quot;</span></span>
<span id="cb30-26"><a href="#cb30-26" tabindex="-1"></a>)</span>
<span id="cb30-27"><a href="#cb30-27" tabindex="-1"></a></span>
<span id="cb30-28"><a href="#cb30-28" tabindex="-1"></a><span class="co"># Define color and linetype mapping</span></span>
<span id="cb30-29"><a href="#cb30-29" tabindex="-1"></a>color_mapping <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb30-30"><a href="#cb30-30" tabindex="-1"></a>  <span class="st">&quot;Statnu0.5&quot;</span> <span class="ot">=</span> <span class="st">&quot;blue&quot;</span>, </span>
<span id="cb30-31"><a href="#cb30-31" tabindex="-1"></a>  <span class="st">&quot;Nonstatnu0.5&quot;</span> <span class="ot">=</span> <span class="st">&quot;red&quot;</span></span>
<span id="cb30-32"><a href="#cb30-32" tabindex="-1"></a>)</span>
<span id="cb30-33"><a href="#cb30-33" tabindex="-1"></a></span>
<span id="cb30-34"><a href="#cb30-34" tabindex="-1"></a>linetype_mapping <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb30-35"><a href="#cb30-35" tabindex="-1"></a>  <span class="st">&quot;Statnu0.5&quot;</span> <span class="ot">=</span> <span class="st">&quot;dotdash&quot;</span>, </span>
<span id="cb30-36"><a href="#cb30-36" tabindex="-1"></a>  <span class="st">&quot;Nonstatnu0.5&quot;</span> <span class="ot">=</span> <span class="st">&quot;solid&quot;</span></span>
<span id="cb30-37"><a href="#cb30-37" tabindex="-1"></a>)</span>
<span id="cb30-38"><a href="#cb30-38" tabindex="-1"></a></span>
<span id="cb30-39"><a href="#cb30-39" tabindex="-1"></a><span class="co"># Plot MSE</span></span>
<span id="cb30-40"><a href="#cb30-40" tabindex="-1"></a>mse_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(mse_long, <span class="fu">aes</span>(<span class="at">x =</span> distance, <span class="at">y =</span> MSE, <span class="at">color =</span> nu, <span class="at">linetype =</span> nu)) <span class="sc">+</span></span>
<span id="cb30-41"><a href="#cb30-41" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb30-42"><a href="#cb30-42" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;MSE&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Distance&quot;</span>) <span class="sc">+</span></span>
<span id="cb30-43"><a href="#cb30-43" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> color_mapping, <span class="at">labels =</span> label_mapping, <span class="at">name =</span> <span class="fu">expression</span>(nu)) <span class="sc">+</span></span>
<span id="cb30-44"><a href="#cb30-44" tabindex="-1"></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values =</span> linetype_mapping, <span class="at">labels =</span> label_mapping, <span class="at">name =</span> <span class="fu">expression</span>(nu)) <span class="sc">+</span></span>
<span id="cb30-45"><a href="#cb30-45" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb30-46"><a href="#cb30-46" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span>
<span id="cb30-47"><a href="#cb30-47" tabindex="-1"></a></span>
<span id="cb30-48"><a href="#cb30-48" tabindex="-1"></a><span class="co"># Plot negative log-score</span></span>
<span id="cb30-49"><a href="#cb30-49" tabindex="-1"></a>ls_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(ls_long, <span class="fu">aes</span>(<span class="at">x =</span> distance, <span class="at">y =</span> LogScore, <span class="at">color =</span> nu, <span class="at">linetype =</span> nu)) <span class="sc">+</span></span>
<span id="cb30-50"><a href="#cb30-50" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb30-51"><a href="#cb30-51" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;Negative Log-Score&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Distance&quot;</span>) <span class="sc">+</span></span>
<span id="cb30-52"><a href="#cb30-52" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> color_mapping, <span class="at">labels =</span> label_mapping, <span class="at">name =</span> <span class="fu">expression</span>(nu)) <span class="sc">+</span></span>
<span id="cb30-53"><a href="#cb30-53" tabindex="-1"></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values =</span> linetype_mapping, <span class="at">labels =</span> label_mapping, <span class="at">name =</span> <span class="fu">expression</span>(nu)) <span class="sc">+</span></span>
<span id="cb30-54"><a href="#cb30-54" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb30-55"><a href="#cb30-55" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text =</span> <span class="fu">element_text</span>(<span class="at">family =</span> <span class="st">&quot;Palatino&quot;</span>))</span>
<span id="cb30-56"><a href="#cb30-56" tabindex="-1"></a></span>
<span id="cb30-57"><a href="#cb30-57" tabindex="-1"></a><span class="co"># Combine plots with a shared legend at the top in a single line</span></span>
<span id="cb30-58"><a href="#cb30-58" tabindex="-1"></a>combined_plot <span class="ot">&lt;-</span> mse_plot <span class="sc">+</span> ls_plot <span class="sc">+</span> </span>
<span id="cb30-59"><a href="#cb30-59" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">&#39;collect&#39;</span>) <span class="sc">&amp;</span> </span>
<span id="cb30-60"><a href="#cb30-60" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&#39;right&#39;</span>) <span class="sc">&amp;</span> </span>
<span id="cb30-61"><a href="#cb30-61" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">nrow =</span> <span class="dv">2</span>), <span class="at">linetype =</span> <span class="fu">guide_legend</span>(<span class="at">nrow =</span> <span class="dv">1</span>))</span>
<span id="cb30-62"><a href="#cb30-62" tabindex="-1"></a></span>
<span id="cb30-63"><a href="#cb30-63" tabindex="-1"></a><span class="co"># Display combined plot</span></span>
<span id="cb30-64"><a href="#cb30-64" tabindex="-1"></a><span class="fu">print</span>(combined_plot)</span></code></pre></div>
<p><img src="non_stationarity_files/figure-html/unnamed-chunk-22-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="references" class="section level1" number="5">
<h1><span class="header-section-number">5</span> References</h1>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="fu">cite_packages</span>(<span class="at">output =</span> <span class="st">&quot;paragraph&quot;</span>, <span class="at">out.dir =</span> <span class="st">&quot;.&quot;</span>)</span></code></pre></div>
<p>We used R version 4.4.0 <span class="citation">(R Core Team
2024)</span> and the following R packages: ggtext v. 0.1.2 <span
class="citation">(Wilke and Wiernik 2022)</span>, here v. 1.0.1 <span
class="citation">(Müller 2020)</span>, htmltools v. 0.5.8.1 <span
class="citation">(Cheng et al. 2024)</span>, INLA v. 24.6.27 <span
class="citation">(Rue, Martino, and Chopin 2009; Lindgren, Rue, and
Lindström 2011; Martins et al. 2013; Lindgren and Rue 2015; De Coninck
et al. 2016; Rue et al. 2017; Verbosio et al. 2017; Bakka et al. 2018;
Kourounis, Fuchs, and Schenk 2018)</span>, inlabru v. 2.10.1.9010 <span
class="citation">(Yuan et al. 2017; Bachl et al. 2019)</span>, knitr v.
1.47 <span class="citation">(Xie 2014, 2015, 2024)</span>, latex2exp v.
0.9.6 <span class="citation">(Meschiari 2022)</span>, listviewer v.
4.0.0 <span class="citation">(de Jong, Gainer, and Russell 2023)</span>,
mapview v. 2.11.2 <span class="citation">(Appelhans et al. 2023)</span>,
Matrix v. 1.6.5 <span class="citation">(Bates, Maechler, and Jagan
2024)</span>, MetricGraph v. 1.3.0.9000 <span class="citation">(Bolin,
Simas, and Wallin 2023b, 2023a, 2023c, 2024; Bolin et al. 2023)</span>,
patchwork v. 1.2.0 <span class="citation">(Pedersen 2024)</span>, plotly
v. 4.10.4 <span class="citation">(Sievert 2020)</span>, plotrix v. 3.8.4
<span class="citation">(J 2006)</span>, reshape2 v. 1.4.4 <span
class="citation">(Wickham 2007)</span>, rmarkdown v. 2.27 <span
class="citation">(Xie, Allaire, and Grolemund 2018; Xie, Dervieux, and
Riederer 2020; Allaire et al. 2024)</span>, rSPDE v. 2.3.3.9000 <span
class="citation">(Bolin and Kirchner 2020; Bolin and Simas 2023; Bolin,
Simas, and Xiong 2023)</span>, scales v. 1.3.0 <span
class="citation">(Wickham, Pedersen, and Seidel 2023)</span>, sf v.
1.0.16 <span class="citation">(Pebesma 2018; Pebesma and Bivand
2023)</span>, tidyverse v. 2.0.0 <span class="citation">(Wickham et al.
2019)</span>, TSstudio v. 0.1.7 <span class="citation">(Krispin
2023)</span>, xaringanExtra v. 0.8.0 <span class="citation">(Aden-Buie
and Warkentin 2024)</span>.</p>
<div id="refs" class="references csl-bib-body hanging-indent"
entry-spacing="0">
<div id="ref-xaringanExtra" class="csl-entry">
Aden-Buie, Garrick, and Matthew T. Warkentin. 2024. <em><span
class="nocase">xaringanExtra</span>: Extras and Extensions for
<span>“<span class="nocase">xaringan</span>”</span> Slides</em>. <a
href="https://CRAN.R-project.org/package=xaringanExtra">https://CRAN.R-project.org/package=xaringanExtra</a>.
</div>
<div id="ref-rmarkdown2024" class="csl-entry">
Allaire, JJ, Yihui Xie, Christophe Dervieux, Jonathan McPherson, Javier
Luraschi, Kevin Ushey, Aron Atkins, et al. 2024. <em><span
class="nocase">rmarkdown</span>: Dynamic Documents for r</em>. <a
href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>.
</div>
<div id="ref-mapview" class="csl-entry">
Appelhans, Tim, Florian Detsch, Christoph Reudenbach, and Stefan
Woellauer. 2023. <em><span class="nocase">mapview</span>: Interactive
Viewing of Spatial Data in r</em>. <a
href="https://CRAN.R-project.org/package=mapview">https://CRAN.R-project.org/package=mapview</a>.
</div>
<div id="ref-inlabru2019" class="csl-entry">
Bachl, Fabian E., Finn Lindgren, David L. Borchers, and Janine B.
Illian. 2019. <span>“<span class="nocase">inlabru</span>: An
<span>R</span> Package for <span>Bayesian</span> Spatial Modelling from
Ecological Survey Data.”</span> <em>Methods in Ecology and
Evolution</em> 10: 760–66. <a
href="https://doi.org/10.1111/2041-210X.13168">https://doi.org/10.1111/2041-210X.13168</a>.
</div>
<div id="ref-INLA2018f" class="csl-entry">
Bakka, Haakon, Håvard Rue, Geir-Arne Fuglstad, Andrea I. Riebler, David
Bolin, Janine Illian, Elias Krainski, Daniel P. Simpson, and Finn K.
Lindgren. 2018. <span>“Spatial Modelling with <span>INLA</span>:
<span>A</span> Review.”</span> <em>WIRES (Invited Extended Review)</em>
xx (Feb): xx–. <a
href="http://arxiv.org/abs/1802.06350">http://arxiv.org/abs/1802.06350</a>.
</div>
<div id="ref-Matrix" class="csl-entry">
Bates, Douglas, Martin Maechler, and Mikael Jagan. 2024.
<em><span>Matrix</span>: Sparse and Dense Matrix Classes and
Methods</em>. <a
href="https://CRAN.R-project.org/package=Matrix">https://CRAN.R-project.org/package=Matrix</a>.
</div>
<div id="ref-rSPDE2020c" class="csl-entry">
Bolin, David, and Kristin Kirchner. 2020. <span>“The Rational
<span>SPDE</span> Approach for <span>Gaussian</span> Random Fields with
General Smoothness.”</span> <em>Journal of Computational and Graphical
Statistics</em> 29 (2): 274–85. <a
href="https://doi.org/10.1080/10618600.2019.1665537">https://doi.org/10.1080/10618600.2019.1665537</a>.
</div>
<div id="ref-MetricGraph2023c" class="csl-entry">
Bolin, David, Mihály Kovács, Vivek Kumar, and Alexandre B. Simas. 2023.
<span>“Regularity and Numerical Approximation of Fractional Elliptic
Differential Equations on Compact Metric Graphs.”</span> <em>Mathematics
of Computation</em>. <a
href="https://doi.org/10.1090/mcom/3929">https://doi.org/10.1090/mcom/3929</a>.
</div>
<div id="ref-rSPDE2023a" class="csl-entry">
Bolin, David, and Alexandre B. Simas. 2023. <em><span
class="nocase">rSPDE</span>: Rational Approximations of Fractional
Stochastic Partial Differential Equations</em>. <a
href="https://CRAN.R-project.org/package=rSPDE">https://CRAN.R-project.org/package=rSPDE</a>.
</div>
<div id="ref-MetricGraph2023d" class="csl-entry">
Bolin, David, Alexandre B. Simas, and Jonas Wallin. 2023a. <span>“Markov
Properties of Gaussian Random Fields on Compact Metric Graphs.”</span>
<em>arXiv Preprint arXiv:2304.03190</em>. <a
href="https://doi.org/10.48550/arXiv.2304.03190">https://doi.org/10.48550/arXiv.2304.03190</a>.
</div>
<div id="ref-MetricGraph2023a" class="csl-entry">
———. 2023b. <em><span>MetricGraph</span>: Random Fields on Metric
Graphs</em>. <a
href="https://CRAN.R-project.org/package=MetricGraph">https://CRAN.R-project.org/package=MetricGraph</a>.
</div>
<div id="ref-MetricGraph2023e" class="csl-entry">
———. 2023c. <span>“Statistical Inference for Gaussian Whittle-Matérn
Fields on Metric Graphs.”</span> <em>arXiv Preprint
arXiv:2304.10372</em>. <a
href="https://doi.org/10.48550/arXiv.2304.10372">https://doi.org/10.48550/arXiv.2304.10372</a>.
</div>
<div id="ref-MetricGraph2024b" class="csl-entry">
———. 2024. <span>“Gaussian Whittle-Matérn Fields on Metric
Graphs.”</span> <em>Bernoulli</em> 30 (2): 1611–39. <a
href="https://doi.org/10.3150/23-BEJ1647">https://doi.org/10.3150/23-BEJ1647</a>.
</div>
<div id="ref-rSPDE2023b" class="csl-entry">
Bolin, David, Alexandre B. Simas, and Zhen Xiong. 2023.
<span>“Covariance-Based Rational Approximations of Fractional SPDEs for
Computationally Efficient Bayesian Inference.”</span> <em>Journal of
Computational and Graphical Statistics</em>. <a
href="https://doi.org/10.1080/10618600.2023.2231051">https://doi.org/10.1080/10618600.2023.2231051</a>.
</div>
<div id="ref-htmltools" class="csl-entry">
Cheng, Joe, Carson Sievert, Barret Schloerke, Winston Chang, Yihui Xie,
and Jeff Allen. 2024. <em><span class="nocase">htmltools</span>: Tools
for HTML</em>. <a
href="https://CRAN.R-project.org/package=htmltools">https://CRAN.R-project.org/package=htmltools</a>.
</div>
<div id="ref-INLA2016g" class="csl-entry">
De Coninck, Arne, Bernard De Baets, Drosos Kourounis, Fabio Verbosio,
Olaf Schenk, Steven Maenhout, and Jan Fostier. 2016.
<span>“<span>Needles</span>: Toward Large-Scale Genomic Prediction with
Marker-by-Environment Interaction.”</span> <em>Genetics</em> 203 (1):
543–55. <a
href="https://doi.org/10.1534/genetics.115.179887">https://doi.org/10.1534/genetics.115.179887</a>.
</div>
<div id="ref-listviewer" class="csl-entry">
de Jong, Jos, Mac Gainer, and Kent Russell. 2023. <em><span
class="nocase">listviewer</span>: <span>“<span
class="nocase">htmlwidget</span>”</span> for Interactive Views of r
Lists</em>. <a
href="https://CRAN.R-project.org/package=listviewer">https://CRAN.R-project.org/package=listviewer</a>.
</div>
<div id="ref-plotrix" class="csl-entry">
J, Lemon. 2006. <span>“<span>Plotrix</span>: A Package in the Red Light
District of r.”</span> <em>R-News</em> 6 (4): 8–12.
</div>
<div id="ref-INLA2018i" class="csl-entry">
Kourounis, D., A. Fuchs, and O. Schenk. 2018. <span>“Towards the Next
Generation of Multiperiod Optimal Power Flow Solvers.”</span> <em>IEEE
Transactions on Power Systems</em> PP (99): 1–10. <a
href="https://doi.org/10.1109/TPWRS.2017.2789187">https://doi.org/10.1109/TPWRS.2017.2789187</a>.
</div>
<div id="ref-TSstudio" class="csl-entry">
Krispin, Rami. 2023. <em><span>TSstudio</span>: Functions for Time
Series Analysis and Forecasting</em>. <a
href="https://CRAN.R-project.org/package=TSstudio">https://CRAN.R-project.org/package=TSstudio</a>.
</div>
<div id="ref-INLA2015d" class="csl-entry">
Lindgren, Finn, and Håvard Rue. 2015. <span>“Bayesian Spatial Modelling
with <span>R</span>-<span>INLA</span>.”</span> <em>Journal of
Statistical Software</em> 63 (19): 1–25. <a
href="http://www.jstatsoft.org/v63/i19/">http://www.jstatsoft.org/v63/i19/</a>.
</div>
<div id="ref-INLA2011c" class="csl-entry">
Lindgren, Finn, Håvard Rue, and Johan Lindström. 2011. <span>“An
Explicit Link Between <span>Gaussian</span> Fields and
<span>Gaussian</span> <span>Markov</span> Random Fields: The Stochastic
Partial Differential Equation Approach (with Discussion).”</span>
<em>Journal of the Royal Statistical Society B</em> 73 (4): 423–98.
</div>
<div id="ref-INLA2013b" class="csl-entry">
Martins, Thiago G., Daniel Simpson, Finn Lindgren, and Håvard Rue. 2013.
<span>“Bayesian Computing with <span>INLA</span>: <span>N</span>ew
Features.”</span> <em>Computational Statistics and Data Analysis</em>
67: 68–83.
</div>
<div id="ref-latex2exp" class="csl-entry">
Meschiari, Stefano. 2022. <em>Latex2exp: Use LaTeX Expressions in
Plots</em>. <a
href="https://CRAN.R-project.org/package=latex2exp">https://CRAN.R-project.org/package=latex2exp</a>.
</div>
<div id="ref-here" class="csl-entry">
Müller, Kirill. 2020. <em><span class="nocase">here</span>: A Simpler
Way to Find Your Files</em>. <a
href="https://CRAN.R-project.org/package=here">https://CRAN.R-project.org/package=here</a>.
</div>
<div id="ref-sf2018" class="csl-entry">
Pebesma, Edzer. 2018. <span>“<span class="nocase">Simple Features for R:
Standardized Support for Spatial Vector Data</span>.”</span>
<em><span>The R Journal</span></em> 10 (1): 439–46. <a
href="https://doi.org/10.32614/RJ-2018-009">https://doi.org/10.32614/RJ-2018-009</a>.
</div>
<div id="ref-sf2023" class="csl-entry">
Pebesma, Edzer, and Roger Bivand. 2023. <em><span class="nocase">Spatial
Data Science: With applications in R</span></em>. <span>Chapman and
Hall/CRC</span>. <a
href="https://doi.org/10.1201/9780429459016">https://doi.org/10.1201/9780429459016</a>.
</div>
<div id="ref-patchwork" class="csl-entry">
Pedersen, Thomas Lin. 2024. <em><span class="nocase">patchwork</span>:
The Composer of Plots</em>. <a
href="https://CRAN.R-project.org/package=patchwork">https://CRAN.R-project.org/package=patchwork</a>.
</div>
<div id="ref-base" class="csl-entry">
R Core Team. 2024. <em><span>R</span>: A Language and Environment for
Statistical Computing</em>. Vienna, Austria: R Foundation for
Statistical Computing. <a
href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-INLA2009a" class="csl-entry">
Rue, Håvard, Sara Martino, and Nicholas Chopin. 2009. <span>“Approximate
<span>Bayesian</span> Inference for Latent <span>Gaussian</span> Models
Using Integrated Nested <span>Laplace</span> Approximations (with
Discussion).”</span> <em>Journal of the Royal Statistical Society B</em>
71: 319–92.
</div>
<div id="ref-INLA2017e" class="csl-entry">
Rue, Håvard, Andrea I. Riebler, Sigrunn H. Sørbye, Janine B. Illian,
Daniel P. Simpson, and Finn K. Lindgren. 2017. <span>“Bayesian Computing
with <span>INLA</span>: <span>A</span> Review.”</span> <em>Annual
Reviews of Statistics and Its Applications</em> 4 (March): 395–421. <a
href="http://arxiv.org/abs/1604.00860">http://arxiv.org/abs/1604.00860</a>.
</div>
<div id="ref-plotly" class="csl-entry">
Sievert, Carson. 2020. <em>Interactive Web-Based Data Visualization with
r, Plotly, and Shiny</em>. Chapman; Hall/CRC. <a
href="https://plotly-r.com">https://plotly-r.com</a>.
</div>
<div id="ref-INLA2017h" class="csl-entry">
Verbosio, Fabio, Arne De Coninck, Drosos Kourounis, and Olaf Schenk.
2017. <span>“Enhancing the Scalability of Selected Inversion
Factorization Algorithms in Genomic Prediction.”</span> <em>Journal of
Computational Science</em> 22 (Supplement C): 99–108. <a
href="https://doi.org/10.1016/j.jocs.2017.08.013">https://doi.org/10.1016/j.jocs.2017.08.013</a>.
</div>
<div id="ref-reshape2" class="csl-entry">
Wickham, Hadley. 2007. <span>“Reshaping Data with the <span
class="nocase">reshape</span> Package.”</span> <em>Journal of
Statistical Software</em> 21 (12): 1–20. <a
href="http://www.jstatsoft.org/v21/i12/">http://www.jstatsoft.org/v21/i12/</a>.
</div>
<div id="ref-tidyverse" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy
D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019.
<span>“Welcome to the <span class="nocase">tidyverse</span>.”</span>
<em>Journal of Open Source Software</em> 4 (43): 1686. <a
href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
<div id="ref-scales" class="csl-entry">
Wickham, Hadley, Thomas Lin Pedersen, and Dana Seidel. 2023. <em><span
class="nocase">scales</span>: Scale Functions for Visualization</em>. <a
href="https://CRAN.R-project.org/package=scales">https://CRAN.R-project.org/package=scales</a>.
</div>
<div id="ref-ggtext" class="csl-entry">
Wilke, Claus O., and Brenton M. Wiernik. 2022. <em><span
class="nocase">ggtext</span>: Improved Text Rendering Support for
<span>“<span class="nocase">ggplot2</span>”</span></em>. <a
href="https://CRAN.R-project.org/package=ggtext">https://CRAN.R-project.org/package=ggtext</a>.
</div>
<div id="ref-knitr2014" class="csl-entry">
Xie, Yihui. 2014. <span>“<span class="nocase">knitr</span>: A
Comprehensive Tool for Reproducible Research in <span>R</span>.”</span>
In <em>Implementing Reproducible Computational Research</em>, edited by
Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman;
Hall/CRC.
</div>
<div id="ref-knitr2015" class="csl-entry">
———. 2015. <em>Dynamic Documents with <span>R</span> and Knitr</em>. 2nd
ed. Boca Raton, Florida: Chapman; Hall/CRC. <a
href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-knitr2024" class="csl-entry">
———. 2024. <em><span class="nocase">knitr</span>: A General-Purpose
Package for Dynamic Report Generation in r</em>. <a
href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</div>
<div id="ref-rmarkdown2018" class="csl-entry">
Xie, Yihui, J. J. Allaire, and Garrett Grolemund. 2018. <em>R Markdown:
The Definitive Guide</em>. Boca Raton, Florida: Chapman; Hall/CRC. <a
href="https://bookdown.org/yihui/rmarkdown">https://bookdown.org/yihui/rmarkdown</a>.
</div>
<div id="ref-rmarkdown2020" class="csl-entry">
Xie, Yihui, Christophe Dervieux, and Emily Riederer. 2020. <em>R
Markdown Cookbook</em>. Boca Raton, Florida: Chapman; Hall/CRC. <a
href="https://bookdown.org/yihui/rmarkdown-cookbook">https://bookdown.org/yihui/rmarkdown-cookbook</a>.
</div>
<div id="ref-inlabru2017" class="csl-entry">
Yuan, Yuan, Bachl, Fabian E., Lindgren, Finn, Borchers, et al. 2017.
<span>“Point Process Models for Spatio-Temporal Distance Sampling Data
from a Large-Scale Survey of Blue Whales.”</span> <em>Ann. Appl.
Stat.</em> 11 (4): 2270–97. <a
href="https://doi.org/10.1214/17-AOAS1078">https://doi.org/10.1214/17-AOAS1078</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
